tensorboard/util/timing.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for measuring elapsed time."""
+
+import contextlib
+import logging
+import threading
+import time
+
+from tensorboard.util import tb_logging
+
+logger = tb_logging.get_logger()
+
+
+def log_latency(name_or_function, log_level=None):
+    """Log latency in a function or region.
+
+    This can act as a decorator: if passed a function, it returns a
+    decorated version of the function. Otherwise, the argument is
+    interpreted as a region name, and the result is either a context
+    manager or a decorator at the caller's discretion. Thus, the
```

Comment:
I am open to opinions or suggestions about this design. I considered
making the signaling more in-band by requiring the decorator to be
invoked as `@log_latency()` rather than `@log_latency`, but this has the
downside that if you accidentally invoke it as `@log_latency` then it
does the wrong thing in a confusing way.

One potential alternative would have the following semantics:

  - `log_latency()` returns a decorator;
  - `log_latency(name)` returns a context manager/decorator hybrid when
    `isinstance(name, str)`;
  - `log_latency(x)` raises a `TypeError` when `x` is not a string,
    catching the `@log_latency def foo(): pass` usage error.

Reply:
Okay, I’ve convinced myself: I’ve switched it to the API described in
this comment, and also added tests.